### ISSUE TITLE
Evil makes hidden cursor reappear on window configuration changes

### ISSUE THREAD

**Issue by TheBB** [emacs-evil/evil#592]
Originally reported by: **Vasilij Schneidermann (Bitbucket: [wasamasa](https://bitbucket.org/wasamasa), GitHub: [wasamasa](https://github.com/wasamasa))**

----------------------------------------

I wrote a package that hides the cursor by setting `cursor-type` to `nil`.  This works initially, however changing the window configuration restores it because Evil has `evil-refresh-cursor` in `window-configuration-change-hook`.  The comment above suggests it should just be changing the cursor *color*...

See also [zone-nyan#5](https://github.com/wasamasa/zone-nyan/issues/5).

----------------------------------------
- Bitbucket: https://bitbucket.org/lyro/evil/issue/592

**Comment by rosetail** [comment]
I'm not sure if this will help anyone, but I found a way to hide the cursor in `evil mode`. I tried this as a last-ditch effort and I have now idea how or why this works. I've only tested in `org-agenda` using `setq-local`, so your milage may vary. But before setting `cursor-type` to `nil`, set `evil-default-cursor` to `'(ignore)`.

**Comment by archiif** [comment]
You can also do:
```elisp
(setq-local evil-normal-state-cursor '(bar . 0))
```
You might also need to set other `evil-*-state-cursor` variables, depending on what you want to do.

**Comment by real-or-random** [comment]
```elisp
;; Helpers for hiding the cursor in evil.
(defun >evil-nil-cursor-spec ()
  (setq cursor-type nil))
(defun >evil-hide-cursor-local ()
  "Hide the cursor in the current buffer."
  (interactive)
  (setq-local
   evil-default-cursor #'>evil-nil-cursor-spec
   ;; Add more states here as desired
   evil-normal-state-cursor #'>evil-nil-cursor-spec
   evil-motion-state-cursor #'>evil-nil-cursor-spec))
```

edit: Or, use this much simpler alternative:
```elisp
(setq-local evil-default-cursor (list nil)
            ;; Add more states here as desired
            evil-normal-state-cursor (list nil)
            evil-motion-state-cursor (list nil))
```

**Comment by perryprog** [comment]
@tomdl89 this fix seems to cause the block cursor to be invisible in <del>insert</del> <ins>normal</ins> mode in a bare config (nextstep, macOS). For me I'm seeing `evil-refresh-cursor` first call `evil-set-cursor` with `t`, but the second `evil-set-cursor` call is then `nil` (from `(evil-state-property 'normal :cursor t)`).

**Comment by real-or-random** [comment]
> [@tomdl89](https://github.com/tomdl89) this fix seems to cause the block cursor to be invisible in insert mode in a bare config (nextstep, macOS). For me I'm seeing `evil-refresh-cursor` first call `evil-set-cursor` with `t`, but the second `evil-set-cursor` call is then `nil` (from `(evil-state-property 'normal :cursor t)`).

In insert mode or in normal mode? I see that this would be a problem in normal mode. So far, `:cursor nil` (which is probably implicitly set if no `:cursor` is given) was used to indicate that a cursor should use the default cursor. But now that nil actually means "no cursor", we don't have a value for indicating "use the default cursor". We'd probably need something like `unspecified` in faces.

Maybe it's easier to revert this and simply change the docs to say that `(list nil)` should be passed for "no cursor" (instead of just `nil`).

Sorry I missed this. I use doom Emacs which has explicit settings for all states.

**Comment by perryprog** [comment]
Oops, yes, the cursor is only invisible in normal (and IIRC visual mode)—my mistake.

**Comment by tomdl89** [comment]
I've reverted this. My mistake for not checking before merging. I don't have much time for evil these days but blindly merging PRs is admittedly not a good solution. @real-or-random if you want to make a new PR that actually fixes this properly, and you can thoroughly test it in vanilla evil, not just Doom, then go ahead.

**Comment by real-or-random** [comment]
>  [@real-or-random](https://github.com/real-or-random) if you want to make a new PR that actually fixes this properly, and you can thoroughly test it in vanilla evil, not just Doom, then go ahead.

Sorry for not testing in vanilla Emacs. I didn't expect that this could cause trouble here. 

Given the maturity of the codebase, I tend to think that it's best to simply change the docs to match the actual behavior. This won't require touching actual code and so no testing will be required. See https://github.com/emacs-evil/evil/pull/2026.

**Comment by tomdl89** [comment]
Agreed, that seems like a sensible approach. Thanks @real-or-random !
